examples/fluids/qfunctions/utils_eigensolver_jacobi.h

```diff
@@ -0,0 +1,327 @@
+      switch (sort_criteria) {
+        case SORT_DECREASING_EVALS:
+          if (eval[j] > eval[i_max]) i_max = j;
+          break;
+        case SORT_INCREASING_EVALS:
+          if (eval[j] < eval[i_max]) i_max = j;
+          break;
+        case SORT_DECREASING_ABS_EVALS:
+          if (fabs(eval[j]) > fabs(eval[i_max])) i_max = j;
+          break;
+        case SORT_INCREASING_ABS_EVALS:
+          if (fabs(eval[j]) < fabs(eval[i_max])) i_max = j;
+          break;
+        default:
+          break;
+      }
+    }
+    SwapScalar(&eval[i], &eval[i_max]);
+    for (CeedInt k = 0; k < N; k++) SwapScalar(&evec[i * N + k], &evec[i_max * N + k]);
+  }
+}
+
+/// @brief Calculate all the eigenvalues and eigevectors of a symmetric matrix
+///        using the Jacobi eigenvalue algorithm:
+///        https://en.wikipedia.org/wiki/Jacobi_eigenvalue_algorithm
+/// @returns The number of Jacobi iterations attempted, which should be > 0.
+///          If the return value is not strictly > 0 then convergence failed.
+/// @note  To reduce the computation time further, set calc_evecs=false.
+///        Additionally, note that the output evecs should be normalized. It
+///        simply takes the Identity matrix and performs (isometric) rotations
+///        on it, so divergence from normalized is due to finite-precision
+///        arithmetic of the rotations.
+//
+// @param[in]  A              the matrix you wish to diagonalize (size NxN)
+// @param[in]  N              size of the matrix
+// @param[out] eval           store the eigenvalues here (size N)
+// @param[out] evec           store the eigenvectors here (in rows, size NxN)
+// @param[out] max_idx_row    work vector of size N-1
+// @param[in]  sort_criteria  sort results?
+// @param[in]  calc_evecs     calculate the eigenvectors?
+// @param[in]  max_num_sweeps maximum number of iterations = max_num_sweeps * number of off-diagonals (N*(N-1)/2)
+CEED_QFUNCTION_HELPER CeedInt Diagonalize(CeedScalar *A, CeedInt N, CeedScalar *eval, CeedScalar *evec, CeedInt *max_idx_row,
+                                          SortCriteria sort_criteria, bool calc_evec, const CeedInt max_num_sweeps) {
+  CeedScalar rotmat_cst[3] = {0.};  // cos(θ), sin(θ), and tan(θ),
+
+  if (calc_evec)
+    for (CeedInt i = 0; i < N; i++)
+      for (CeedInt j = 0; j < N; j++) evec[i * N + j] = (i == j) ? 1.0 : 0.0;  // Set evec equal to the identity matrix
+
+  for (CeedInt i = 0; i < N - 1; i++) max_idx_row[i] = MaxEntryRow(A, N, i);
+
+  // -- Iteration --
+  CeedInt n_iters;
+  CeedInt max_num_iters = max_num_sweeps * N * (N - 1) / 2;
+  for (n_iters = 1; n_iters <= max_num_iters; n_iters++) {
+    CeedInt i, j;
+    MaxEntry(A, N, max_idx_row, &i, &j);
+
+    // If A[i][j] is small compared to A[i][i] and A[j][j], set it to 0.
+    if ((A[i * N + i] + A[i * N + j] == A[i * N + i]) && (A[j * N + j] + A[i * N + j] == A[j * N + j])) {
+      A[i * N + j]   = 0.0;
+      max_idx_row[i] = MaxEntryRow(A, N, i);
+    }
```

Comment:
This code currently checks every loop as to the index with the largest entry. I believe this will probably very sub-optimal on GPU. I've left it here for now just to avoid premature optimization, but what I've seen others do is just loop over all the off-diagonal entries (that was for CPU SIMD instructions, but I believe the performance implications are more-or-less the same).

Reply:
Thanks for pointing this out. I suspect it won't be a significant cost for our use case here, but likely will be in solid mechanics where it's in the critical path of the main solver.

Reply:
Gotcha. I feel like it is a significant cost for the model itself, but I guess until we integrate it into the Jacobian it'll only affect the residual evaluation time.